The missing parameter does not accept an IP when executing a terms aggregation on a field of type ip

**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 6.0.0, Build: 8f0685b/2017-11-10T18:41:22.859Z, JVM: 1.8.0_77

**Plugins installed**: []

**JVM version** (`java -version`):
java version "1.8.0_77"
Java(TM) SE Runtime Environment (build 1.8.0_77-b03)
Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
Darwin MacBook-1265.local 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64

**Description of the problem including expected versus actual behavior**:
Providing an IP as the value of the `missing` parameter, when executing a `terms` aggregation on a field of type `ip` results in the following 503 error:

```
{
  "error": {
    "root_cause": [],
    "type": "search_phase_execution_exception",
    "reason": "",
    "phase": "fetch",
    "grouped": true,
    "failed_shards": [],
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "encoded bytes are of incorrect length",
      "caused_by": {
        "type": "unknown_host_exception",
        "reason": "addr is of illegal length"
      }
    }
  },
  "status": 503
}
```
Looking at the code that raises the exception (stack trace pasted below) it seems that the code expects a byte array of length 4 or 16. As a workaround, the IP can be encoded as a character array instead (for example `"\u0000\u0000\u0000\u0000"`).

**Steps to reproduce**:

 1. Create an index with a field mapped as type `ip`:
```
PUT iptest
{
  "mappings": {
    "doc": {
      "properties": {
        "ip": {
          "type": "ip"
        }
      }
    }
  }
}
```
 2. Index a document that is missing the `ip` field:
```
PUT iptest/doc/1
{
  "foo": "bar"
}
```
 3. Execute a `terms` agg, providing an IP address as the value of the `missing` parameter. This results in a 503 error:
```
GET iptest/_search
{
  "size": 0,
  "aggregations": {
    "ip": {
      "terms": {
        "field": "ip",
        "missing": "0.0.0.0"
      }
    }
  }
}
```
4. Pass a 4-character string as the value of `missing`. This is executed successfully:
```
GET iptest/_search
{
  "size": 0,
  "aggregations": {
    "ip": {
      "terms": {
        "field": "ip",
        "missing": "\u0000\u0000\u0000\u0000"
      }
    }
  }
}
```

**Provide logs (if relevant)**:
```
[2017-12-13T09:12:41,550][WARN ][r.suppressed             ] path: /iptest/_search, params: {index=iptest}
org.elasticsearch.action.search.SearchPhaseExecutionException: 
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:272) [elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase$1.onFailure(FetchSearchPhase.java:92) [elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.onFailure(ThreadContext.java:623) [elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39) [elasticsearch-6.0.0.jar:6.0.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
Caused by: java.lang.IllegalArgumentException: encoded bytes are of incorrect length
	at org.apache.lucene.document.InetAddressPoint.decode(InetAddressPoint.java:187) ~[lucene-misc-7.0.1.jar:7.0.1 8d6c3889aa543954424d8ac1dbb3f03bf207140b - sarowe - 2017-10-02 14:36:35]
	at org.elasticsearch.search.DocValueFormat$4.format(DocValueFormat.java:298) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.StringTerms$Bucket.getKeyAsString(StringTerms.java:75) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.StringTerms$Bucket.getKey(StringTerms.java:64) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.InternalTerms.doReduce(InternalTerms.java:274) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.InternalAggregation.reduce(InternalAggregation.java:120) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.InternalAggregations.reduce(InternalAggregations.java:77) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reduceAggs(SearchPhaseController.java:523) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reducedQueryPhase(SearchPhaseController.java:500) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reducedQueryPhase(SearchPhaseController.java:417) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController$1.reduce(SearchPhaseController.java:736) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:102) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase.access$000(FetchSearchPhase.java:45) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:87) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.0.0.jar:6.0.0]
	... 3 more
Caused by: java.net.UnknownHostException: addr is of illegal length
	at java.net.InetAddress.getByAddress(InetAddress.java:1042) ~[?:1.8.0_77]
	at java.net.InetAddress.getByAddress(InetAddress.java:1439) ~[?:1.8.0_77]
	at org.apache.lucene.document.InetAddressPoint.decode(InetAddressPoint.java:184) ~[lucene-misc-7.0.1.jar:7.0.1 8d6c3889aa543954424d8ac1dbb3f03bf207140b - sarowe - 2017-10-02 14:36:35]
	at org.elasticsearch.search.DocValueFormat$4.format(DocValueFormat.java:298) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.StringTerms$Bucket.getKeyAsString(StringTerms.java:75) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.StringTerms$Bucket.getKey(StringTerms.java:64) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.bucket.terms.InternalTerms.doReduce(InternalTerms.java:274) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.InternalAggregation.reduce(InternalAggregation.java:120) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.search.aggregations.InternalAggregations.reduce(InternalAggregations.java:77) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reduceAggs(SearchPhaseController.java:523) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reducedQueryPhase(SearchPhaseController.java:500) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController.reducedQueryPhase(SearchPhaseController.java:417) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.SearchPhaseController$1.reduce(SearchPhaseController.java:736) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:102) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase.access$000(FetchSearchPhase.java:45) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:87) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) ~[elasticsearch-6.0.0.jar:6.0.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.0.0.jar:6.0.0]
	... 3 more

```

--- a/core/src/main/java/org/elasticsearch/search/DocValueFormat.java
+++ b/core/src/main/java/org/elasticsearch/search/DocValueFormat.java
@@ -56,7 +56,7 @@ public interface DocValueFormat extends NamedWriteable {
      *  such as the {@code long}, {@code double} or {@code date} fields. */
     String format(double value);
 
-    /** Format a double value. This is used by terms aggregations to format
+    /** Format a binary value. This is used by terms aggregations to format
      *  keys for fields that use binary doc value representations such as the
      *  {@code keyword} and {@code ip} fields. */
     String format(BytesRef value);

--- a/core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
@@ -252,7 +252,7 @@ public class ValuesSourceConfig<VS extends ValuesSource> {
         }
 
         if (vs instanceof ValuesSource.Bytes) {
-            final BytesRef missing = new BytesRef(missing().toString());
+            final BytesRef missing = format.parseBytesRef(missing().toString());
             if (vs instanceof ValuesSource.Bytes.WithOrdinals) {
                 return (VS) MissingValues.replaceMissing((ValuesSource.Bytes.WithOrdinals) vs, missing);
             } else {

--- a/core/src/test/java/org/elasticsearch/search/aggregations/bucket/IpTermsIT.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/bucket/IpTermsIT.java
@@ -113,4 +113,29 @@ public class IpTermsIT extends AbstractTermsTestCase {
         assertEquals("2001:db8::2:1", bucket2.getKey());
         assertEquals("2001:db8::2:1", bucket2.getKeyAsString());
     }
+
+    public void testMissingValue() throws Exception {
+        assertAcked(prepareCreate("index").addMapping("type", "ip", "type=ip"));
+        indexRandom(true,
+            client().prepareIndex("index", "type", "1").setSource("ip", "192.168.1.7"),
+            client().prepareIndex("index", "type", "2").setSource("ip", "192.168.1.7"),
+            client().prepareIndex("index", "type", "3").setSource("ip", "127.0.0.1"),
+            client().prepareIndex("index", "type", "4").setSource("not_ip", "something"));
+        SearchResponse response = client().prepareSearch("index").addAggregation(AggregationBuilders
+            .terms("my_terms").field("ip").missing("127.0.0.1").executionHint(randomExecutionHint())).get();
+
+        assertSearchResponse(response);
+        Terms terms = response.getAggregations().get("my_terms");
+        assertEquals(2, terms.getBuckets().size());
+
+        Terms.Bucket bucket1 = terms.getBuckets().get(0);
+        assertEquals(2, bucket1.getDocCount());
+        assertEquals("127.0.0.1", bucket1.getKey());
+        assertEquals("127.0.0.1", bucket1.getKeyAsString());
+
+        Terms.Bucket bucket2 = terms.getBuckets().get(1);
+        assertEquals(2, bucket2.getDocCount());
+        assertEquals("192.168.1.7", bucket2.getKey());
+        assertEquals("192.168.1.7", bucket2.getKeyAsString());
+    }
 }

